README.md
# dropwizard-json-exceptions
Reusable ExceptionMappers for Dropwizard Apps that want more verbose JSON error outputs

Dropwizard intentionally provides pretty obtuse error messages to the end user; if you're interested in providing additional JSON-formatted exceptions for any exceptions that happen during execution of your REST endpoints, then just add this JAR to your classpath and turn off the Dropwizard server's standard exception mappers.

For example, the kinds of responses you'll get from a REST endpoint in your service that encounters a JsonProcessingException might look like:
```
{
    "code": 400,
    "message": "Unable to process JSON",
    "details": "Unexpected token (END_OBJECT), expected FIELD_NAME: missing property 'type' that is to contain type id  (for class com.washingtonpost.arc.ans.v0_3.model.Story)"
}
```

### Dropwizard Compatability Matrix

| Dropwizard Version | Peer-Authenticator Version |
|--------------------|----------------------------|
|        0.8.*       |            1.x.y           |
|        0.9.*       |            N/A             |
|        1.0.*       |            2.x.y           |

## Integration
Add this JAR to your Dropwizard -server's POM

```XML
<dependency>
    <groupId>com.washingtonpost.dropwizard</groupId>
    <artifactId>dropwizard-json-exceptions</artifactId>
    <version>${version.wp.dropwizard.exceptions}</version>
</dependency>
```

Then in your application configuration, just turn off the server default exception mappers:

```YAML
server: 
    registerDefaultExceptionMappers: false
```

In your Application class itself, you can either auto-detect the ExceptionMappers with a guice module configuration like:
```
import com.hubspot.dropwizard.guice.GuiceBundle;

public class MyAppApplication extends Application<MyAppConfiguration> {
    private GuiceBundle<MyAppConfiguration> guiceBundle;

    @Override
    public void initialize(Bootstrap<MyAppConfiguration> bootstrap) {
        guiceBundle = GuiceBundle.<MyAppConfiguration>newBuilder()
                .addModule(new FooModule())
                .enableAutoConfig("com.washingtonpost.dropwizard.exceptions.mappers")
                .setConfigClass(MyAppConfiguration.class)
                .build();
```

Alternatively, you can just manually add whatever exception mappers you want directly to your Jersey ResourceConfig:

```
import com.washingtonpost.dropwizard.exceptions.mappers.JsonProcessingExceptionMapper;
import com.washingtonpost.dropwizard.exceptions.mappers.RuntimeExceptionMapper;

public class MyAppApplication extends Application<MyAppConfiguration> {

    @Override
    public void run(MyAppConfiguration configuration, Environment environment) throws Exception {
        environment.jersey().register(new JsonProcessingExceptionMapper(true));
        environment.jersey().register(new RuntimeExceptionMapper());
        
```

## Configuring Verbosity and Response Formats
The ExceptionMappers in this JAR contain default constructors that create output messages with "details" and in a JSON format.

You can control either of those behaviors by turning details off or providing a MediaType (like MediaType.APPLICATION_JSON_TYPE):
```
    ExceptionMapper noDetailsXmlMapper = 
        new JsonProcessingExceptionMapper(false, MediaType.APPLICATION_XML_TYPE);

```

Note: be mindful of the potential security concerns (information leak) by setting details=true in your runtime exception mapper.


RELEASE_NOTES.md
# dropwizard-json-exceptions

## 2.0.0 2016/12/12

* Upgraded dropwizard dependency to 1.0.5

## 1.1.2 2016/06/21

* All WebAppExceptions should be mapped to JSON with their own response type

## 1.1.1 2016/03/03

* Added 403s to RuntimeExceptionMapper

## 1.1.0 2015/11/19

* Added configurable MediaType return types for the 2 ExceptionMappers (they're still defaulting to JSON)
* Added ErrorMessage JSON to all types of JsonProcessingExceptionMapper responses

## 1.0.0 2015/11/18

* Initial copy of the JsonProcessingExceptionMapper from https://goo.gl/uhbxtQ
* Initial copy of the RuntimeExceptionMapper from http://gary-rowe.com/agilestack/2012/10/23/how-to-implement-a-runtimeexceptionmapper-for-dropwizard/


src/test/java/com/washingtonpost/dropwizard/exceptions/mappers/TestRuntimeExceptionMapper.java
package com.washingtonpost.dropwizard.exceptions.mappers;

import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import org.json.JSONException;
import org.junit.Test;

/**
 * <p>Tests the exception mapping of our RuntimeExceptionMapper</p>
 */
public class TestRuntimeExceptionMapper extends AbstractExceptionMapperTest {

    private static final ExceptionMapper DEFAULT_MAPPER = new RuntimeExceptionMapper();

    @Test
    public void testIllegalStateException() throws JSONException {
        runJsonAssertionTest(DEFAULT_MAPPER,
                             new IllegalStateException("Oops! an ISE!"),
                             500,
                             "illegalStateException-expected.json");

    }

    @Test
    public void testWebAppExceptionUnauthorized() throws JSONException {
        runJsonAssertionTest(DEFAULT_MAPPER,
                             new WebApplicationException("You're not allowed in here!", Response.Status.UNAUTHORIZED),
                             401,
                             "webAppException401-expected.json");
    }

    @Test
    public void testWebAppExceptionNotFound() throws JSONException {
        runJsonAssertionTest(DEFAULT_MAPPER,
                             new WebApplicationException("Can't find it!", Response.Status.NOT_FOUND),
                             404,
                             "webAppException404-expected.json");
    }

    @Test
    public void testWebAppExceptionContentType() throws JSONException {
        runJsonAssertionTest(DEFAULT_MAPPER,
                             new WebApplicationException("Wrong type", Response.Status.UNSUPPORTED_MEDIA_TYPE),
                             415,
                             "webAppException415-expected.json");
    }
}


src/test/java/com/washingtonpost/dropwizard/exceptions/mappers/AbstractExceptionMapperTest.java
package com.washingtonpost.dropwizard.exceptions.mappers;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.xml.XmlMapper;
import io.dropwizard.jersey.errors.ErrorMessage;
import java.io.IOException;
import java.net.URL;
import java.util.Scanner;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import org.custommonkey.xmlunit.XMLAssert;
import org.custommonkey.xmlunit.XMLUnit;
import org.json.JSONException;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.xml.sax.SAXException;

/**
 * <p>Abstract test holding a couple common helper functions</p>
 */
public abstract class AbstractExceptionMapperTest {
    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractExceptionMapperTest.class);
    protected static final ObjectMapper objectMapper = new ObjectMapper();

    /**
     * @param resourcePath The path to the resource to load
     * @return The full URL of the {@code resourcePath}
     */
    protected URL getResource(String resourcePath) {
        return getClass().getClassLoader().getResource(resourcePath);
    }

    /**
     * @param fixtureName The name of the fixture in the same path as the test being executed, e.g. "credit".
     * @return The full URL of a {@code fixtureName} file sitting in the same relative directory as the caller
     */
    protected URL getSisterPathResource(String fixtureName) {
        return getResource(getClass().getPackage().getName().replace(".", "/") + "/" + fixtureName);
    }

    /**
     * @param fixtureName The name of a fixture file to load into a String
     * @return The contents of the fixtureFile, in one big 'ol string
     */
    protected String loadSisterPathFixture(String fixtureName) {
        URL sisterPathResource = getSisterPathResource(fixtureName);
        String fixtureContents;
        try (Scanner scanner = new Scanner(sisterPathResource.openStream(), "UTF-8")) {
            // From "stupid scanner tricks" https://weblogs.java.net/blog/pat/archive/2004/10/stupid_scanner.html
            fixtureContents = scanner.useDelimiter("\\A").next();
            scanner.close();
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
        return fixtureContents;
    }

    protected void runJsonAssertionTest(ExceptionMapper mapper,
                                        Exception exception,
                                        int expectedStatusCode,
                                        String expectedFixture) throws JSONException {
        Response response = mapper.toResponse(exception);
        assertEquals(expectedStatusCode, response.getStatus());
        assertTrue(response.hasEntity());

        String actualJson = readJsonPayload(response.getEntity());
        String expectedJson = loadSisterPathFixture(expectedFixture);

        JSONAssert.assertEquals(expectedJson, actualJson, JSONCompareMode.LENIENT);
    }

    protected String readJsonPayload(Object entity) {
        ErrorMessage errorMessage = (ErrorMessage)entity;
        try {
            String json = objectMapper.writeValueAsString(errorMessage);
            return json;
        } catch (JsonProcessingException ex) {
            fail("Unable to parse JSON out of entity " + errorMessage);
        }
        return null;
    }

    protected void runXmlAssertionTest(ExceptionMapper mapper,
                                       Exception exception,
                                       int expectedStatusCode,
                                       String expectedFixture) throws JSONException, IOException, SAXException {
        Response response = mapper.toResponse(exception);
        assertEquals(expectedStatusCode, response.getStatus());
        assertTrue(response.hasEntity());

        String actualXML = readXmlPayload(response.getEntity());
        LOGGER.debug("actualXML = {}", actualXML);


        String expectedXML = loadSisterPathFixture(expectedFixture);

        XMLUnit.setIgnoreWhitespace(true);
        XMLAssert.assertXMLEqual(expectedXML, actualXML);
    }

    protected String readXmlPayload(Object entity) throws IOException {
        ErrorMessage errorMessage = (ErrorMessage)entity;
        XmlMapper xmlMapper = new XmlMapper();
        return xmlMapper.writeValueAsString(errorMessage);
    }

}


src/test/java/com/washingtonpost/dropwizard/exceptions/mappers/TestJsonProcessingExceptionMapper.java
package com.washingtonpost.dropwizard.exceptions.mappers;

import com.fasterxml.jackson.core.JsonGenerationException;
import com.fasterxml.jackson.core.JsonProcessingException;
import java.io.IOException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.ext.ExceptionMapper;
import org.json.JSONException;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.xml.sax.SAXException;

/**
 * <p>Tests the exception mapping of our JsonProcessor</p>
 */
public class TestJsonProcessingExceptionMapper extends AbstractExceptionMapperTest {
    private static final Logger LOGGER = LoggerFactory.getLogger(JsonProcessingExceptionMapper.class);
    private static final ExceptionMapper DEFAULT_JSON_MAPPER = new JsonProcessingExceptionMapper();

    @Test
    public void testJsonGenerationException() throws JsonProcessingException, JSONException {
        runJsonAssertionTest(DEFAULT_JSON_MAPPER,
                new JsonGenerationException("There was a problem"),
                500,
                "jsonGenerationException-expected.json");
    }

    @Test
    public void testJsonProcessingNoSuitableConstructor() throws JSONException {
        runJsonAssertionTest(DEFAULT_JSON_MAPPER,
                new MyJPE("No suitable constructor found for blah"),
                500,
                "jsonProcessingNoSuitableConstructor-expected.json");
    }

    @Test
    public void testJsonProcessingException() throws JSONException {
        runJsonAssertionTest(DEFAULT_JSON_MAPPER,
                new MyJPE("Something went wrong here!"),
                400,
                "jsonProcessingException-expected.json");
    }

    /**
     * If we don't construct our ExceptionMapper with a "true" flag for whether or not to show details,
     * we should get a slightly less verbose output
     * @throws org.json.JSONException
     */
    @Test
    public void testJsonProcessingExceptionWithNoDetailsMapper() throws JSONException {
        ExceptionMapper mapper = new JsonProcessingExceptionMapper(false, MediaType.APPLICATION_JSON_TYPE);
        runJsonAssertionTest(mapper,
                new MyJPE("Something went wrong here!"),
                400,
                "jsonProcessingExceptionNoDetails-expected.json");
    }

    @Test
    public void testJsonProcessingExceptionWithXMLOutputType() throws JSONException, IOException, SAXException {
        ExceptionMapper mapper = new JsonProcessingExceptionMapper(true, MediaType.APPLICATION_XML_TYPE);
        runXmlAssertionTest(mapper,
                            new MyJPE("Something went wrong here!"),
                            400,
                            "testJsonProcessingExceptionWithXMLOutputType-expected.xml");
    }

    // Cuz the actual JsonProcessingException is all protected
    private class MyJPE extends JsonProcessingException {
        public MyJPE(String message) {
            super(message);
        }
    }
}


src/main/java/com/washingtonpost/dropwizard/exceptions/mappers/JsonProcessingExceptionMapper.java
package com.washingtonpost.dropwizard.exceptions.mappers;

import com.fasterxml.jackson.core.JsonGenerationException;
import com.fasterxml.jackson.core.JsonProcessingException;
import io.dropwizard.jersey.errors.ErrorMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

/**
 * <p>Tells Dropwizard how to send a Response when our code throws a JsonProcessingException</p>
 * <p>From https://goo.gl/uhbxtQ</p>
 */
@Provider
public class JsonProcessingExceptionMapper implements ExceptionMapper<JsonProcessingException> {
    private static final Logger LOGGER = LoggerFactory.getLogger(JsonProcessingExceptionMapper.class);
    private static final boolean DEFAULT_SHOW_DETAILS = true;
    private static final MediaType DEFAULT_MEDIA_TYPE = MediaType.APPLICATION_JSON_TYPE;
    private final boolean showDetails;
    private final MediaType mediaType;

    public JsonProcessingExceptionMapper() {
        this(DEFAULT_SHOW_DETAILS, DEFAULT_MEDIA_TYPE);
    }

    public JsonProcessingExceptionMapper(boolean showDetails, MediaType mediaType) {
        this.showDetails = showDetails;
        this.mediaType = mediaType;
    }

    @Override
    public Response toResponse(JsonProcessingException exception) {
        /*
         * If the error is in the JSON generation, it's a server error.
         */
        if (exception instanceof JsonGenerationException) {
            LOGGER.warn("Error generating JSON", exception);
            return Response
                    .serverError()
                    .type(this.mediaType)
                    .entity(new ErrorMessage(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(),
                            "Error generating JSON",
                            showDetails? exception.getOriginalMessage() : null))
                    .build();
        }

        final String message = exception.getOriginalMessage();

        /*
         * If we can't deserialize the JSON because someone forgot a no-arg constructor, it's a
         * server error and we should inform the developer.
         */
        if (message.startsWith("No suitable constructor found")) {
            LOGGER.error("Unable to deserialize the specific type", exception);
            return Response
                    .serverError()
                    .type(this.mediaType)
                    .entity(new ErrorMessage(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(),
                            "Unable to deserialize the specific type",
                            showDetails? exception.getOriginalMessage() : null))
                    .build();
        }

        /*
         * Otherwise, it's those pesky users.
         */
        LOGGER.debug("Unable to process JSON", exception);
        final ErrorMessage errorMessage = new ErrorMessage(Response.Status.BAD_REQUEST.getStatusCode(),
                "Unable to process JSON", showDetails ? message : null);
        return Response.status(Response.Status.BAD_REQUEST)
                .type(this.mediaType)
                .entity(errorMessage)
                .build();
    }
}


src/main/java/com/washingtonpost/dropwizard/exceptions/mappers/RuntimeExceptionMapper.java
package com.washingtonpost.dropwizard.exceptions.mappers;

import io.dropwizard.jersey.errors.ErrorMessage;
import java.io.PrintWriter;
import java.io.StringWriter;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * <p>
 * From http://gary-rowe.com/agilestack/2012/10/23/how-to-implement-a-runtimeexceptionmapper-for-dropwizard/</p>
 * <p>
 * Provider to provide the following to Jersey framework:</p>
 * <ul>
 * <li>Provision of general runtime exception to response mapping</li>
 * </ul>
 * <p>The RuntimeExceptionMapper has a DEFAULT_SHOW_DETAILS set to false because it's questionable security to emit
 * stack traces from a REST API and it makes upgrading and testing the exception mapper hard because comparing against stack
 * trace Strings for equality or near-equality is pretty lame</p>
 */
@Provider
public class RuntimeExceptionMapper implements ExceptionMapper<RuntimeException> {

    private static final Logger LOGGER = LoggerFactory.getLogger(RuntimeExceptionMapper.class);
    private static final boolean DEFAULT_SHOW_DETAILS = false;
    private static final MediaType DEFAUT_MEDIA_TYPE = MediaType.APPLICATION_JSON_TYPE;
    private final boolean showDetails;
    private final MediaType mediaType;

    public RuntimeExceptionMapper() {
        this(DEFAULT_SHOW_DETAILS, DEFAUT_MEDIA_TYPE);
    }

    public RuntimeExceptionMapper(boolean showDetails, MediaType mediaType) {
        this.showDetails = showDetails;
        this.mediaType = mediaType;
    }

    @Override
    public Response toResponse(RuntimeException exception) {

        // Build default response
        Response defaultResponse = Response.serverError()
                .type(this.mediaType)
                .entity(new ErrorMessage(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(),
                                         exception.getMessage(),
                                         this.showDetails ? stackTraceToString(exception) : null))
                .build();

        // Check for any specific handling
        if (exception instanceof WebApplicationException) {
            LOGGER.info("Using specific handler");
            return handleWebApplicationException(exception, defaultResponse);
        }

        // Use the default
        LOGGER.error(exception.getMessage(), exception);
        return defaultResponse;
    }

    private Response handleWebApplicationException(RuntimeException exception, Response defaultResponse) {
        WebApplicationException webAppException = (WebApplicationException) exception;

        // No logging
        if (webAppException.getResponse().getStatus() == 401) {
            return Response
                    .status(Response.Status.UNAUTHORIZED)
                    .type(this.mediaType)
                    .entity(new ErrorMessage(Response.Status.UNAUTHORIZED.getStatusCode(),
                                             exception.getMessage(),
                                             this.showDetails ? stackTraceToString(exception) : null))
                    .build();
        }

        if (webAppException.getResponse().getStatus() == 403) {
            return Response
                    .status(Response.Status.FORBIDDEN)
                    .type(this.mediaType)
                    .entity(new ErrorMessage(Response.Status.FORBIDDEN.getStatusCode(),
                                             exception.getMessage(),
                                             this.showDetails ? stackTraceToString(exception) : null))
                    .build();
        }


        if (webAppException.getResponse().getStatus() == 404) {
            return Response
                    .status(Response.Status.NOT_FOUND)
                    .type(this.mediaType)
                    .entity(new ErrorMessage(Response.Status.NOT_FOUND.getStatusCode(),
                                             exception.getMessage(),
                                             this.showDetails ? stackTraceToString(exception) : null))
                    .build();
        }

        LOGGER.error(exception.getMessage(), exception);

        Response r = Response
            .status(webAppException.getResponse().getStatus())
            .type(this.mediaType)
            .entity(new ErrorMessage(webAppException.getResponse().getStatus(),
                                     exception.getMessage(),
                                     this.showDetails ? stackTraceToString(exception) : null))
            .build();

        return r;

    }

    // Basically the same as org.apache.commons.lang.exception.ExceptionUtils.getStackTrace(Throwable), without the dependency
    private String stackTraceToString(Throwable throwable) {
        StringWriter stringWriter = new StringWriter();
        PrintWriter printWriter = new PrintWriter(stringWriter);
        throwable.printStackTrace(printWriter);
        return stringWriter.toString();
    }
}
